File: app/src/main/java/com/productivity/habits/ui/analytics/MonthlyHeatmapGrid.kt
package com.productivity.habits.ui.analytics

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowLeft
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.productivity.habits.ui.common.HapticsHelper
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.DateTimeFormatter

data class HeatmapDayData(
    val date: LocalDate,
    val completedCount: Int,
    val scheduledCount: Int,
    val ratePercent: Int
)

@Composable
fun MonthlyHeatmapGrid(
    month: YearMonth,
    dayDataMap: Map<LocalDate, HeatmapDayData>,
    onPreviousMonth: () -> Unit,
    onNextMonth: () -> Unit,
    modifier: Modifier = Modifier
) {
    val haptic = LocalHapticFeedback.current
    var selectedDayDetail by remember { mutableStateOf<HeatmapDayData?>(null) }

    val daysInMonth = month.lengthOfMonth()
    val firstDayOfMonth = month.atDay(1)
    val dayOfWeekOffset = firstDayOfMonth.dayOfWeek.value % 7

    if (selectedDayDetail != null) {
        val detail = selectedDayDetail!!
        AlertDialog(
            onDismissRequest = { selectedDayDetail = null },
            title = {
                Text(
                    text = detail.date.format(DateTimeFormatter.ofPattern("EEEE, MMM d, yyyy")),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold
                )
            },
            text = {
                Column {
                    Text("Completion Rate: ${detail.ratePercent}%", fontWeight = FontWeight.SemiBold)
                    Spacer(modifier = Modifier.height(4.dp))
                    Text("Completed Habits: ${detail.completedCount} of ${detail.scheduledCount}")
                }
            },
            confirmButton = {
                TextButton(onClick = { selectedDayDetail = null }) {
                    Text("Close")
                }
            }
        )
    }

    Column(modifier = modifier.fillMaxWidth()) {
        // Month Navigation Header
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = month.format(DateTimeFormatter.ofPattern("MMMM yyyy")),
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold
            )

            Row(verticalAlignment = Alignment.CenterVertically) {
                IconButton(
                    onClick = {
                        HapticsHelper.performLightHaptic(haptic)
                        onPreviousMonth()
                    },
                    modifier = Modifier.size(32.dp)
                ) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.KeyboardArrowLeft,
                        contentDescription = "Previous Month"
                    )
                }
                IconButton(
                    onClick = {
                        HapticsHelper.performLightHaptic(haptic)
                        onNextMonth()
                    },
                    modifier = Modifier.size(32.dp)
                ) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.KeyboardArrowRight,
                        contentDescription = "Next Month"
                    )
                }
            }
        }

        Spacer(modifier = Modifier.height(12.dp))

        // Heatmap Grid
        val totalCells = ((dayOfWeekOffset + daysInMonth + 6) / 7) * 7
        val primary = MaterialTheme.colorScheme.primary

        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            for (row in 0 until (totalCells / 7)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    for (col in 0 until 7) {
                        val cellIndex = (row * 7) + col
                        val dayNumber = cellIndex - dayOfWeekOffset + 1

                        if (dayNumber in 1..daysInMonth) {
                            val date = month.atDay(dayNumber)
                            val dayData = dayDataMap[date] ?: HeatmapDayData(date, 0, 0, 0)
                            val rate = dayData.ratePercent

                            val cellColor = when {
                                rate == 100 -> primary
                                rate >= 50 -> primary.copy(alpha = 0.65f)
                                rate > 0 -> primary.copy(alpha = 0.3f)
                                else -> MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)
                            }

                            Box(
                                modifier = Modifier
                                    .weight(1f)
                                    .aspectRatio(1f)
                                    .padding(2.dp)
                                    .clip(RoundedCornerShape(6.dp))
                                    .background(cellColor)
                                    .clickable {
                                        HapticsHelper.performLightHaptic(haptic)
                                        selectedDayDetail = dayData
                                    },
                                contentAlignment = Alignment.Center
                            ) {
                                Text(
                                    text = "$dayNumber",
                                    style = MaterialTheme.typography.labelSmall,
                                    fontWeight = if (rate >= 50) FontWeight.Bold else FontWeight.Normal,
                                    color = if (rate >= 50) Color.White else MaterialTheme.colorScheme.onSurfaceVariant
                                )
                            }
                        } else {
                            Spacer(modifier = Modifier.weight(1f))
                        }
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(14.dp))

        // Heatmap Legend
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.End,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "Less",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(modifier = Modifier.width(6.dp))

            listOf(
                MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f),
                primary.copy(alpha = 0.3f),
                primary.copy(alpha = 0.65f),
                primary
            ).forEach { color ->
                Box(
                    modifier = Modifier
                        .size(12.dp)
                        .padding(horizontal = 1.dp)
                        .clip(RoundedCornerShape(2.dp))
                        .background(color)
                )
            }

            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = "More",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

File: lib/ui/analytics/widgets/monthly_heatmap_grid.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../../common/haptics_helper.dart';
import '../controllers/analytics_controller.dart';

class MonthlyHeatmapGrid extends StatefulWidget {
  final DateTime month;
  final Map<DateTime, HeatmapDayData> dayDataMap;
  final VoidCallback onPreviousMonth;
  final VoidCallback onNextMonth;

  const MonthlyHeatmapGrid({
    super.key,
    required this.month,
    required this.dayDataMap,
    required this.onPreviousMonth,
    required this.onNextMonth,
  });

  @override
  State<MonthlyHeatmapGrid> createState() => _MonthlyHeatmapGridState();
}

class _MonthlyHeatmapGridState extends State<MonthlyHeatmapGrid> {
  void _showDayDetailDialog(BuildContext context, HeatmapDayData detail) {
    final theme = Theme.of(context);
    final dateStr =
        DateFormat('EEEE, MMM d, yyyy').format(detail.date);

    showDialog(
      context: context,
      builder: (context) {
        return AlertDialog(
          title: Text(
            dateStr,
            style: theme.textTheme.titleMedium?.copyWith(
              fontWeight: FontWeight.bold,
            ),
          ),
          content: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                'Completion Rate: ${detail.ratePercent}%',
                style: const TextStyle(fontWeight: FontWeight.w600),
              ),
              const SizedBox(height: 4),
              Text(
                'Completed Habits: ${detail.completedCount} of ${detail.scheduledCount}',
              ),
            ],
          ),
          actions: [
            TextButton(
              onPressed: () => Navigator.of(context).pop(),
              child: const Text('Close'),
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final primary = theme.colorScheme.primary;

    final daysInMonth =
        DateTime(widget.month.year, widget.month.month + 1, 0).day;
    final firstDayOfMonth =
        DateTime(widget.month.year, widget.month.month, 1);
    // Sunday is 0 in Dart (weekday % 7: Monday is 1, Sunday is 7%7=0)
    final dayOfWeekOffset = firstDayOfMonth.weekday % 7;
    final totalCells =
        (((dayOfWeekOffset + daysInMonth + 6) ~/ 7) * 7);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        // Month Navigation Header
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(
              DateFormat('MMMM yyyy').format(widget.month),
              style: theme.textTheme.titleMedium?.copyWith(
                fontWeight: FontWeight.bold,
              ),
            ),
            Row(
              children: [
                IconButton(
                  icon: const Icon(Icons.keyboard_arrow_left),
                  tooltip: 'Previous Month',
                  iconSize: 20,
                  onPressed: () {
                    HapticsHelper.performLightHaptic();
                    widget.onPreviousMonth();
                  },
                ),
                IconButton(
                  icon: const Icon(Icons.keyboard_arrow_right),
                  tooltip: 'Next Month',
                  iconSize: 20,
                  onPressed: () {
                    HapticsHelper.performLightHaptic();
                    widget.onNextMonth();
                  },
                ),
              ],
            ),
          ],
        ),

        const SizedBox(height: 12),

        // Heatmap Grid
        Column(
          children: List.generate(totalCells ~/ 7, (row) {
            return Padding(
              padding: const EdgeInsets.symmetric(vertical: 2),
              child: Row(
                children: List.generate(7, (col) {
                  final cellIndex = (row * 7) + col;
                  final dayNumber = cellIndex - dayOfWeekOffset + 1;

                  if (dayNumber >= 1 && dayNumber <= daysInMonth) {
                    final date = DateTime(
                      widget.month.year,
                      widget.month.month,
                      dayNumber,
                    );
                    final dayData = widget.dayDataMap[date] ??
                        HeatmapDayData(
                          date: date,
                          completedCount: 0,
                          scheduledCount: 0,
                          ratePercent: 0,
                        );
                    final rate = dayData.ratePercent;

                    final Color cellColor;
                    if (rate == 100) {
                      cellColor = primary;
                    } else if (rate >= 50) {
                      cellColor = primary.withValues(alpha: 0.65);
                    } else if (rate > 0) {
                      cellColor = primary.withValues(alpha: 0.3);
                    } else {
                      cellColor = theme.colorScheme.surfaceContainerHighest
                          .withValues(alpha: 0.4);
                    }

                    return Expanded(
                      child: Padding(
                        padding: const EdgeInsets.all(2),
                        child: AspectRatio(
                          aspectRatio: 1,
                          child: InkWell(
                            onTap: () {
                              HapticsHelper.performLightHaptic();
                              _showDayDetailDialog(context, dayData);
                            },
                            borderRadius: BorderRadius.circular(6),
                            child: Container(
                              decoration: BoxDecoration(
                                color: cellColor,
                                borderRadius: BorderRadius.circular(6),
                              ),
                              child: Center(
                                child: Text(
                                  '$dayNumber',
                                  style: theme.textTheme.labelSmall?.copyWith(
                                    fontSize: 10,
                                    fontWeight: rate >= 50
                                        ? FontWeight.bold
                                        : FontWeight.normal,
                                    color: rate >= 50
                                        ? Colors.white
                                        : theme.colorScheme.onSurfaceVariant,
                                  ),
                                ),
                              ),
                            ),
                          ),
                        ),
                      ),
                    );
                  } else {
                    return const Expanded(
                      child: Padding(
                        padding: EdgeInsets.all(2),
                        child: AspectRatio(aspectRatio: 1),
                      ),
                    );
                  }
                }),
              ),
            );
          }),
        ),

        const SizedBox(height: 14),

        // Heatmap Legend
        Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            Text(
              'Less',
              style: theme.textTheme.labelSmall?.copyWith(
                color: theme.colorScheme.onSurfaceVariant,
              ),
            ),
            const SizedBox(width: 6),
            ...[
              theme.colorScheme.surfaceContainerHighest.withValues(alpha: 0.4),
              primary.withValues(alpha: 0.3),
              primary.withValues(alpha: 0.65),
              primary,
            ].map((color) {
              return Container(
                width: 12,
                height: 12,
                margin: const EdgeInsets.symmetric(horizontal: 1.5),
                decoration: BoxDecoration(
                  color: color,
                  borderRadius: BorderRadius.circular(2),
                ),
              );
            }),
            const SizedBox(width: 6),
            Text(
              'More',
              style: theme.textTheme.labelSmall?.copyWith(
                color: theme.colorScheme.onSurfaceVariant,
              ),
            ),
          ],
        ),
      ],
    );
  }
}
